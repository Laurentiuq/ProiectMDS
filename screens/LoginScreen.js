import { KeyboardAvoidingView, StyleSheet, Text, View, TextInput, TouchableOpacity } from 'react-native'
import React, {useEffect, useState} from 'react'
import auth from '../firebase'
import { useNavigation } from '@react-navigation/native';
import { getAuth, createUserWithEmailAndPassword, signInWithEmailAndPassword } from 'firebase/auth'
export default function LoginScreen() {
    const [email, setEmail] = useState('')
    const [password, setPassword] = useState('')


    // get the navigation object from useNavigation hook
    const navigation = useNavigation();


    // Redirect to HomeScreen if user logged in
    useEffect(() => {
        const auth = getAuth();
        const unsubcribe = 
        auth.onAuthStateChanged((user) => {
            if (user) {
                navigation.replace('Home');
            }
        })
        return unsubcribe;
    }, [])


    // This is used in onpress event of signup button
    const handleSignUp = () => {
        const auth = getAuth();
        createUserWithEmailAndPassword(auth, email, password)
        .then((userCredential) => {
            const user = userCredential.user;
            console.log('Registered with', user.email);
        })
        .catch((error) => {
            const errorCode = error.code;
            const errorMessage = error.message;
            console.log(errorCode, errorMessage);
        });
    }

    // GO to register screen when press register button
    const goToRegister = () => {
        navigation.navigate('Register');
    }



    // This is used in onpress event of login button
    const handleLogin = () => {
        const auth = getAuth();
        signInWithEmailAndPassword(auth, email, password)
        .then((userCredential) => {
            const user = userCredential.user;
            console.log('Logged in with', user.email);
        })
        .catch((error) => {
            const errorCode = error.code;
            const errorMessage = error.message;
            console.log(errorCode, errorMessage);
        });
    }
    const handlePaswordReset = () => {
        navigation.navigate('ForgotPassword');
    }

    const handlePasswordReset2 = () => {
        const auth = getAuth();
        auth.sendPasswordResetEmail(email)
        .then(() => {
            console.log('Password reset email sent');
        }
        )
        .catch((error) => {
            const errorCode = error.code;
            const errorMessage = error.message;
            console.log(errorCode, errorMessage);
        });
    }

  return (
    <View
    style={styles.container}
    behavior="padding">
    <View style={styles.inputContainer}>
        <TextInput
            placeholder="Email"
            value = {email}
            // This is used to update the state of email so that when user types in email, it is updated in state
            onChangeText = {text => setEmail(text)}
            style={styles.input}
        />
         <TextInput
            placeholder="Pasword"
            value = {password}
            // This is used to update the state of password
            onChangeText = {text => setPassword(text)} 
            style={styles.input}
            secureTextEntry
        />
    </View>
    <View style={styles.buttonContainer}>
        <TouchableOpacity
          onPress = {handleLogin}
          style = {styles.button}>
            <Text style={styles.buttonText}>Login</Text>
        </TouchableOpacity>
    </View>
    <View style={styles.buttonContainer}>
        <TouchableOpacity
          onPress = {goToRegister}
          style = {[styles.button, styles.buttonOutline]}>
            <Text style={styles.buttonOutlineText}>Register</Text>
        </TouchableOpacity>
    </View>
    <View style={styles.buttonContainer}>
        <TouchableOpacity
          onPress = {handlePaswordReset}
          style = {[styles.button, styles.buttonOutline]}>
            <Text style={styles.buttonOutlineText}>Forgot password</Text>
        </TouchableOpacity>
    </View>
    </View>
  )
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        backgroundColor: 'rgb(227, 150, 150)'
    },
    inputContainer: {
        width: '80%'
    },
    input: {
        backgroundColor: 'rgb(249, 224, 221)',
        paddingHorizontal: 15,
        paddingVertical: 10,
        borderRadius: 10,
        marginTop:5
    },
    buttonContainer: {
        width: '60%',
        justifyContent: 'center',
        alignItmes: 'center',
        marginTop: 20,

    },
    button: {
        backgroundColor: 'rgb(241, 105, 86)',
        width: '100%',
        padding: 15,
        borderRadius: 10,
        alignItems: 'center',
    },
    buttonOutline: {
        backgroundColor: 'white',
        marginTop: 5,
        borderColor: 'rgb(241, 105, 86)',
        borderWidth: 2,
    },
    buttonText:  {
        color:'white',
        fontWeight: '800',
        fontSize: 16,
    },
    buttonOutlineText: {
        color:'rgb(241, 105, 86)',
        fontWeight: '800',
        fontSize: 16,
    },
    
})