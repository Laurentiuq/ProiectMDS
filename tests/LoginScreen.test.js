import { render, fireEvent, waitFor } from '@testing-library/react-native';
import LoginScreen from '../screens/LoginScreen'
;
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

jest.mock('firebase/auth', () => {
    return {
        getAuth: jest.fn(() => ({
            onAuthStateChanged: jest.fn(),
        })),
        signInWithEmailAndPassword: jest.fn(() => Promise.resolve('user')),
    };
});

describe('LoginScreen', () => {
    it('renders correctly', () => {
        const { toJSON } = render(<LoginScreen />);
        expect(toJSON()).toMatchSnapshot();
    });

    it('allows entering an email', () => {
        const { getByPlaceholderText } = render(<LoginScreen />);
        const input = getByPlaceholderText('Introduce your email');
        fireEvent.changeText(input, 'test@test.com');
        expect(input.props.value).toBe('test@test.com');
    });

    it('allows entering a password', () => {
        const { getByPlaceholderText } = render(<LoginScreen />);
        const input = getByPlaceholderText('Introduce your password');
        fireEvent.changeText(input, 'password123');
        expect(input.props.value).toBe('password123');
    });

    it('shows error for invalid email', async () => {
        const { getByPlaceholderText, getByText } = render(<LoginScreen />);
        const input = getByPlaceholderText('Introduce your email');
        fireEvent.changeText(input, 'test');


        fireEvent.press(getByText('Login'));
        await waitFor(() => {
            expect(getByText('Please enter a valid email address')).toBeDefined();
        });
    });

    it('shows error for invalid password', async () => {
        const { getByPlaceholderText, findByText, getByTestId } = render(<LoginScreen />);
        const emailInput = getByPlaceholderText('Introduce your email');
        fireEvent.changeText(emailInput, 'example@example.com');


        const input = getByPlaceholderText('Introduce your password');
        fireEvent.changeText(input, 'pass');

        fireEvent.press(getByTestId('loginButton'));
        const errorMessage = await findByText('Password must be at least 8 characters');
        expect(errorMessage).toBeDefined();
    });

    it('toggles password visibility', () => {
        const { getByText, getByPlaceholderText } = render(<LoginScreen />);
        const input = getByPlaceholderText('Introduce your password');
        expect(input.props.secureTextEntry).toBe(true);
        fireEvent.press(getByText('Show'));
        expect(input.props.secureTextEntry).toBe(false);
    });

    // it('navigates to ForgotPassword on button press', () => {
    //     const mockNavigation = jest.fn();
    //     const { getByText } = render(<LoginScreen navigation={{ navigate: mockNavigation }} />);
    //     fireEvent.press(getByText('Forgot password'));
    //     expect(mockNavigation).toHaveBeenCalledWith('ForgotPassword');
    // });
});

// npm test -- -u 