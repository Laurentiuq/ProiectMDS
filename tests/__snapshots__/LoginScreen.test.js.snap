// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#E39696",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 20,
    }
  }
>
  <Text
    style={
      {
        "color": "white",
        "fontSize": 30,
        "fontWeight": "bold",
        "marginBottom": 40,
        "textAlign": "center",
      }
    }
  >
    FMIQUIZ
  </Text>
  <Text
    style={
      {
        "fontSize": 20,
        "fontWeight": "bold",
        "marginBottom": 10,
      }
    }
  >
    Email Address
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#F9E0DD",
        "borderRadius": 20,
        "flexDirection": "row",
      }
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/email.png",
        }
      }
      style={
        {
          "height": 24,
          "marginLeft": 32,
          "marginRight": 10,
          "width": 24,
        }
      }
    />
    <TextInput
      autoCompleteType="email"
      keyboardType="email-address"
      onChangeText={[Function]}
      placeholder="Introduce your email"
      style={
        {
          "color": "#F16956",
          "flex": 1,
          "paddingHorizontal": 10,
          "paddingVertical": 10,
        }
      }
      textContentType="emailAddress"
      value=""
    />
  </View>
  <View
    style={
      {
        "height": 24,
      }
    }
  />
  <Text
    style={
      {
        "fontSize": 20,
        "fontWeight": "bold",
        "marginBottom": 10,
      }
    }
  >
    Password
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#F9E0DD",
        "borderRadius": 20,
        "flexDirection": "row",
      }
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/password.png",
        }
      }
      style={
        {
          "height": 24,
          "marginLeft": 32,
          "marginRight": 10,
          "width": 24,
        }
      }
    />
    <TextInput
      onChangeText={[Function]}
      placeholder="Introduce your password"
      secureTextEntry={true}
      style={
        {
          "color": "#F16956",
          "flex": 1,
          "paddingHorizontal": 10,
          "paddingVertical": 10,
        }
      }
      textContentType="password"
      value=""
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
          "padding": 10,
          "paddingEnd": 20,
        }
      }
    >
      <Text>
        Show
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "height": 24,
      }
    }
  />
  <View
    accessibilityState={
      {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#F16956",
        "borderRadius": 20,
        "opacity": 1,
        "padding": 15,
      }
    }
    testID="loginButton"
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
    >
      Login
    </Text>
  </View>
  <View
    style={
      {
        "height": 24,
      }
    }
  />
  <View
    style={
      {
        "alignItems": "center",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "textDecorationLine": "underline",
          }
        }
      >
        Forgot password
      </Text>
    </View>
  </View>
</View>
`;
